lms/models/grouping.py

```diff
@@ -49,13 +55,15 @@ class Type(str, Enum):
     #:
     #: For example if the grouping represents a Canvas section or group then parent_id
     #: will reference the grouping for the course that the section or group belongs to.
-    parent_id = sa.Column(
-        sa.Integer(),
-        sa.ForeignKey("grouping.id", ondelete="cascade"),
-        nullable=True,
-    )
+    parent_id = sa.Column(sa.Integer(), nullable=True)
     children = sa.orm.relationship(
-        "Grouping", backref=sa.orm.backref("parent", remote_side=[id])
+        "Grouping",
+        backref=sa.orm.backref(
+            "parent",
+            remote_side=[id, application_instance_id],
+            overlaps="application_instance",
+        ),
+        overlaps="application_instance",
```

Comment:
The two `overlaps` are necessary to avoid this error from SQLAlchemy:

```
sqlalchemy.exc.InvalidRequestError: One or more mappers failed to initialize - can't proceed with initialization of other mappers. Triggering mapper: 'mapped class Grouping->grouping'. Original exception was: relationship 'Grouping.parent' will copy column grouping.application_instance_id to column grouping.application_instance_id, which conflicts with relationship(s): 'Grouping.application_instance' (copies application_instances.id to grouping.application_instance_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="application_instance"' to the 'Grouping.parent' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
```

The link <https://sqlalche.me/e/14/qzyx> contains some useful explanation.

Reply:
I'm going to freely admit I don't know what this is about, and so can't really comment. I assume it just blew up when you tried it and this was the solution?

Reply:
Yeah, it's something to do with how SQLAlchemy figures out the relationships between tables (or between different rows of the same table, in this instance) in order to hook-up the convenience relationship attributes on the ORM objects like `.parent` and `.children`. For some internal SQLAlchemy reason we need to give it an extra clue in this case